File: Demo/Demo/ViewController.m
//
//  IGRMainViewController.m
//  InstaFilters
//
//  Created by Vitalii Parovishnyk on 12/29/16.
//  Copyright © 2016 IGR Software. All rights reserved.
//

@import IGRFilterCombine;

#import "ViewController.h"
#import "IGRFilterbarCell.h"

@interface ViewController () <UIImagePickerControllerDelegate, UINavigationControllerDelegate,
                              UICollectionViewDelegate, UICollectionViewDataSource,
                              IGRFilterCombineDelegate>

@property (nonatomic, weak  ) IBOutlet UICollectionView *collectionView;
@property (nonatomic, weak  ) IBOutlet UIImageView *imageView;

@property (nonatomic, strong) IGRFilterCombine *filterCombine;

@end

static NSString * const kWorkImageNotification = @"WorkImageNotification";
#define DEMO_IMAGE [UIImage imageNamed:@"demo"]

@implementation ViewController

#pragma mark - Life Cycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.filterCombine = [[IGRFilterCombine alloc] initWithDelegate:self];
    
    NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
    [defaultCenter addObserver:self
                      selector:@selector(setupWorkImage:)
                          name:kWorkImageNotification object:nil];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        [self setupDemoView];
    });
}

- (void)dealloc
{
    NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
    [defaultCenter removeObserver:self name:kWorkImageNotification object:nil];
}

#pragma mark - Private

- (void)setupDemoView
{
    [[NSNotificationCenter defaultCenter] postNotificationName:kWorkImageNotification
                                                        object:DEMO_IMAGE];
}

- (UIImage *)prepareImage
{
    return self.imageView.image;
}

#pragma mark - Action

- (IBAction)onTouchGetImageButton:(UIBarButtonItem *)sender
{
    UIAlertControllerStyle style = UIAlertControllerStyleActionSheet;
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Select image", @"")
                                                                   message:@""
                                                            preferredStyle:style];
    
    alert.popoverPresentationController.barButtonItem = sender;
    
    __weak typeof(self) weak = self;
    void(^completeActionBlock)(UIImagePickerControllerSourceType) = ^(UIImagePickerControllerSourceType type) {
        UIImagePickerController *pickerView = [[UIImagePickerController alloc] init];
        pickerView.delegate = weak;
        [pickerView setSourceType:type];
        [self presentViewController:pickerView animated:YES completion:nil];
    };
    
    UIAlertAction *action = [UIAlertAction actionWithTitle:NSLocalizedString(@"From Library", @"")
                                                     style:UIAlertActionStyleDefault
                                                   handler:^(UIAlertAction * action) {
                                                       completeActionBlock(UIImagePickerControllerSourceTypePhotoLibrary);
                                                   }];
    [alert addAction:action];
    
    UIAlertAction *cameraAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"From Camera", @"")
                                                           style:UIAlertActionStyleDefault
                                                         handler:^(UIAlertAction * action) {
                                                             completeActionBlock(UIImagePickerControllerSourceTypeCamera);
                                                         }];
    [alert addAction:cameraAction];
    
    UIAlertAction* cancel = [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"")
                                                     style:UIAlertActionStyleCancel
                                                   handler:nil];
				
    [alert addAction:cancel];
    
    [self presentViewController:alert animated:YES completion:nil];
}

- (IBAction)onTouchShareButton:(UIBarButtonItem *)sender
{
    UIImage *image = [self prepareImage];
    
    UIActivityViewController *avc = [[UIActivityViewController alloc] initWithActivityItems:@[image]
                                                                      applicationActivities:nil];
    avc.popoverPresentationController.barButtonItem = sender;
    avc.popoverPresentationController.permittedArrowDirections = UIPopoverArrowDirectionUp;
    
    [self presentViewController:avc animated:YES completion:nil];
}

#pragma mark - UICollectionViewDataSource

- (NSInteger)collectionView:(UICollectionView *)collectionView
     numberOfItemsInSection:(NSInteger)section
{
    return self.filterCombine.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView
                  cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    IGRFilterbarCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"IGRFilterbarCell"
                                                                       forIndexPath:indexPath];
    
    cell.icon.image = [self.filterCombine filteredPreviewImageAtIndex:indexPath.row];
    cell.title.text = [self.filterCombine filtereNameAtIndex:indexPath.row];
    
    return cell;
}

#pragma mark - UICollectionViewDelegate

- (void)collectionView:(UICollectionView *)collectionView
didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    CGFloat borderOffset = [self igr_borderOffsetFromFiltersbar:collectionView];
    
    UICollectionViewLayoutAttributes *attrs = [collectionView layoutAttributesForItemAtIndexPath:indexPath];
    CGRect frame = attrs.frame;
    frame = UIEdgeInsetsInsetRect(frame, UIEdgeInsetsMake(0.0, -borderOffset, 0.0, -borderOffset));
    [collectionView scrollRectToVisible:frame animated:YES];
    
    self.imageView.image = [self.filterCombine filteredImageAtIndex:indexPath.row];
}

- (void)collectionView:(UICollectionView *)collectionView
didDeselectItemAtIndexPath:(NSIndexPath *)indexPath
{
    UICollectionViewCell *cell = [collectionView cellForItemAtIndexPath:indexPath];
    [cell setSelected:NO];
}

- (CGSize)collectionView:(UICollectionView *)collectionView
                  layout:(nonnull UICollectionViewLayout *)collectionViewLayout
  sizeForItemAtIndexPath:(nonnull NSIndexPath *)indexPath
{
    return CGSizeMake(100.0, 100.0);
}

- (CGFloat)igr_borderOffsetFromFiltersbar:(UICollectionView *)collectionView
{
    return collectionView.frame.size.height * 0.5;
}

#pragma mark - IGRFilterCombineDelegate

- (CGSize)previewSize
{
    return CGSizeMake(70.0, 70.0);
}

#pragma mark - NSNotificationCenter

- (void)setupWorkImage:(NSNotification *)notification
{
    NSAssert([notification.object isKindOfClass:[UIImage class]], @"Image only allowed!");
    UIImage *image = notification.object;
    
    __weak typeof(self) weak = self;
    [self.filterCombine setImage:image
                      completion:^(UIImage * _Nullable processedImage, NSUInteger idx) {
                          NSIndexPath *indexPath = [NSIndexPath indexPathForRow:idx inSection:0];
                          if ([[weak.collectionView indexPathsForSelectedItems] containsObject:indexPath])
                          {
                              dispatch_async(dispatch_get_main_queue(), ^{
                                  weak.imageView.image = processedImage;
                              });
                          }
                      }
                         preview:^(UIImage * _Nullable processedImage, NSUInteger idx) {
                             NSIndexPath *indexPath = [NSIndexPath indexPathForRow:idx inSection:0];
                             if ([[weak.collectionView indexPathsForVisibleItems] containsObject:indexPath])
                             {
                                 dispatch_async(dispatch_get_main_queue(), ^{
                                     IGRFilterbarCell *cell = (IGRFilterbarCell *)[weak.collectionView cellForItemAtIndexPath:indexPath];
                                     cell.icon.image = processedImage;
                                 });
                             }
                         }];
    
    self.imageView.image = image;
    [self.collectionView reloadData];
}

#pragma mark - PickerDelegates

- (void)imagePickerController:(UIImagePickerController *)picker
didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info
{
    UIImage *img = [info valueForKey:UIImagePickerControllerOriginalImage];
    [[NSNotificationCenter defaultCenter] postNotificationName:kWorkImageNotification
                                                        object:img];
    
    [picker dismissViewControllerAnimated:YES completion:nil];
}

@end

File: Demo/DemoSwift/ViewController.swift
//
//  ViewController.swift
//  DemoSwift
//
//  Created by Vitalii Parovishnyk on 1/8/17.
//  Copyright © 2017 IGR Software. All rights reserved.
//

import UIKit

import IGRFilterCombine

class ViewController: UIViewController {
    
    @IBOutlet weak fileprivate var collectionView: UICollectionView?
    @IBOutlet weak fileprivate var imageView: UIImageView?
    
    fileprivate var filterCombine: IGRFilterCombine?
    
    static let kImageNotification = NSNotification.Name(rawValue: "WorkImageNotification")
    let kDemoImage = UIImage.init(named: "demo")
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        filterCombine = IGRFilterCombine(delegate: self as IGRFilterCombineDelegate)
        
        let notificationCenter = NotificationCenter.default
        notificationCenter.addObserver(self,
                                       selector: #selector(setupWorkImage(_:)),
                                       name: ViewController.kImageNotification,
                                       object: nil)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        DispatchQueue.once(token: "com.igrsoft.fastfilter.demo") {
            self.setupDemoView()
        }
    }
    
    deinit {
        let notificationCenter = NotificationCenter.default
        notificationCenter.removeObserver(self,
                                          name: ViewController.kImageNotification,
                                          object: nil)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    fileprivate func setupDemoView() {
        let notificationCenter = NotificationCenter.default
        notificationCenter.post(name: ViewController.kImageNotification, object: kDemoImage)
    }
    
    func setupWorkImage(_ notification: Notification) {
        assert(notification.object is UIImage, "Image only allowed!")
        let image = notification.object as! UIImage
        
        self.filterCombine?.setImage(image, completion: { (processedImage, idx) in
            let indexPath = IndexPath(row: Int(idx), section: 0)
            if (self.collectionView?.indexPathsForSelectedItems?.contains(indexPath))! {
                self.imageView?.image = processedImage
            }
        }) { (processedImage, idx) in
            let indexPath = IndexPath(row: Int(idx), section: 0)
            if (self.collectionView?.indexPathsForVisibleItems.contains(indexPath))! {
                let cell = self.collectionView?.cellForItem(at: indexPath) as! IGRFilterbarCell
                cell.icon?.image = processedImage
            }
        }
        
        self.imageView?.image = image
        self.collectionView?.reloadData()
    }
    
    func prepareImage() -> UIImage {
        return (self.imageView?.image)!
    }
    
    @IBAction func onTouchGetImageButton(_ sender: UIBarButtonItem)
    {
        let alert = UIAlertController.init(title: NSLocalizedString("Select image", comment: ""), message: "", preferredStyle: UIAlertControllerStyle.actionSheet)
        
        alert.popoverPresentationController?.barButtonItem = sender
        
        func completeAlert(type: UIImagePickerControllerSourceType) {
            let pickerView = UIImagePickerController.init()
            pickerView.delegate = self
            pickerView.sourceType = type
            self.present(pickerView, animated: true, completion: nil)
        }
        
        let style = UIAlertActionStyle.default
        var action = UIAlertAction.init(title: NSLocalizedString("From Library", comment: ""),
                                        style: style) { (UIAlertAction) in
                                            completeAlert(type: UIImagePickerControllerSourceType.photoLibrary)
        }
        
        alert.addAction(action)
        
        action = UIAlertAction.init(title: NSLocalizedString("From Camera", comment: ""),
                                    style: style) { (UIAlertAction) in
                                        completeAlert(type: UIImagePickerControllerSourceType.camera)
        }
        
        alert.addAction(action)
        
        action = UIAlertAction.init(title: NSLocalizedString("Cancel", comment: ""),
                                    style: style)
        
        alert.addAction(action)
        
        self.present(alert, animated: true, completion: nil)
    }
    
    
    @IBAction func onTouchShareButton(_ sender: UIBarButtonItem) {
        
        let image = prepareImage()
        
        let avc = UIActivityViewController.init(activityItems: [image], applicationActivities: nil)
        
        avc.popoverPresentationController?.barButtonItem = sender
        avc.popoverPresentationController?.permittedArrowDirections = .up
        
        self.present(avc, animated: true, completion: nil)
    }
    
    func igr_borderOffsetFromFiltersbar(collectionView: UICollectionView) -> CGFloat {
        return collectionView.frame.size.height * 0.5
    }
}

extension ViewController : UICollectionViewDelegate, UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return Int(self.filterCombine!.count())
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "IGRFilterbarCell", for: indexPath) as! IGRFilterbarCell
        
        cell.icon?.image = self.filterCombine?.filteredPreviewImage(at: UInt(indexPath.row))
        cell.title?.text = self.filterCombine?.filtereName(at: UInt(indexPath.row))
        
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return CGSize(width: 100.0, height: 100.0)
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let borderOffset = self.igr_borderOffsetFromFiltersbar(collectionView: collectionView)
        
        let attrs = collectionView.layoutAttributesForItem(at: indexPath)
        var frame = attrs?.frame
        frame = UIEdgeInsetsInsetRect(frame!, UIEdgeInsetsMake(0.0, -borderOffset, 0.0, -borderOffset))
        collectionView.scrollRectToVisible(frame!, animated: true)
        
        self.imageView?.image = self.filterCombine?.filteredImage(at: UInt(indexPath.row))
    }
}

extension ViewController : UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        let image = info[UIImagePickerControllerOriginalImage]
        let notificationCenter = NotificationCenter.default
        notificationCenter.post(name: ViewController.kImageNotification, object: image)
        
        picker.dismiss(animated: true, completion: nil)
    }
}

// MARK: - IGRFilterCombineDelegate

extension ViewController : IGRFilterCombineDelegate {
    func previewSize() -> CGSize {
        return CGSize(width: 70.0, height: 70.0)
    }
}

public extension DispatchQueue {
    
    private static var _onceTracker = [String]()
    
    /**
     Executes a block of code, associated with a unique token, only once.  The code is thread safe and will
     only execute the code once even in the presence of multithreaded calls.
     
     - parameter token: A unique reverse DNS style name such as com.vectorform.<name> or a GUID
     - parameter block: Block to execute once
     */
    public class func once(token: String, block:(Void)->Void) {
        objc_sync_enter(self); defer { objc_sync_exit(self) }
        
        if _onceTracker.contains(token) {
            return
        }
        
        _onceTracker.append(token)
        block()
    }
}

